Lib/statistics.py

```diff
@@ -373,15 +393,20 @@ def median(data):
     4.0
 
     """
-    data = sorted(data)
+    data = [d for d in data]
     n = len(data)
     if n == 0:
         raise StatisticsError("no median for empty data")
-    if n%2 == 1:
-        return data[n//2]
+
+    if n == 1:
+        return data[0]
+    if n == 2:
+        return sum(data) / 2
+    pos = n//2
+    if n%2 != 0:
+        return _quick_select(data, 0, n-1, pos)
     else:
-        i = n//2
-        return (data[i - 1] + data[i])/2
+        return _quick_select(data, 0, n-1, pos-1)/2 + _quick_select(data, 0, n-1, pos)/2
```

Comment:
I'd be surprised if there wasn't a way to avoid doing the whole expensive operation twice.